### PR TITLE
TrajCluster and ClusterAna module improvements

### PR DESCRIPTION
**Pull request by brballer**

### CODE REVIEW

**Comment by gartung** [comment]
code-checks

**Comment by gartung** [comment]
trigger build

**Comment by gartung** [comment]
Sorry for the spurious messages. I am trying to get the attachments to comments woerking.

**Comment by gartung** [comment]
In this case "attach" file means make a comment from the contents of the file. The -R report file and -m "your message here" are mutually exclusive.

**Comment by brballer** [comment]
I expected that there would be problems and am ready to help.

Bruce

**Comment by brballer** [comment]
What is the next step?

**Comment by gartung** [comment]
At least one of the people in the team @LArSoft/level-2-managers needs to do a code review and indicate that they approve the pull request.

**Comment by gartung** [comment]
code-checks

**Comment by knoepfel** [comment]
@brballer, I'm happy to look over the changes, but this PR will not be merged into this week's release.  I'll give it top priority tomorrow.

File: larreco/RecoAlg/TCAlg/Utils.cxx [hunk 1]
```diff
@@ -2544,6 +2705,28 @@ namespace tca {
     return atan2(dw, dt);
   } // TwoTPAngle
 
+  ////////////////////////////////////////////////
+  std::vector<unsigned int> PutHitsInVector(TCSlice& slc, PFPStruct const& pfp, HitStatus_t hitRequest)
```
Comment (knoepfel) [inline larreco/RecoAlg/TCAlg/Utils.cxx]: This is the perfect kind of function: it receives arguments, doesn't appear to modify them, and then creates an object and returns it.  Such a function is naturally thread-safe.  Is it possible to mark `slc` as `TCSlice const&` as `slc` doesn't appear to be modified?

**Comment by brballer** [comment]
Kyle,

I can do that. Similar changes can be made elsewhere in the code.

Bruce

**Comment by knoepfel** [comment]
@brballer, I've looked over the PR.  In general, it's clear to me that the coding changes you made improve what was there before.  I have a few questions:

- Is the intention that `ClusterAnaV2` eventually replace `ClusterAna`?  Or are they intended to be separate?  My primary concern is having to maintain two modules that do similar things.  If the two modules are supposed to exist together, then I would perhaps remain `ClusterAnaV2` to be something like `MultiTPCClusterAna`.  If `ClusterAnaV2` will eventually replace `ClusterAna`, then we should come up with a migration plan.
- A lot of commented-out code has been left in there.  Is that because you intend to work further on it?  Or is it just code you explored and decided not to use?  I ask because there are changes in lardata coming in the nearish future that will require changing some function calls...and we'll need to know which code will to adjust.  If you intend the commented-out code to be worked on later, could you place a comment at the top of the section that says so?  Otherwise, just removing it will make it easier for us to handle.

Two very minor things, neither of which would hold up approval:
- `std::cout` is used in many places.  If you use (e.g.) `mf::LogInfo(...)` instead, then users can direct the output wherever they wish.  Otherwise the output is always logged to STDOUT even if the user doesn't want it to be.  It's possible to script a change to use `mf::LogInfo` instead, but I understand that be a route you wish to not pursue.
- The C++ way of specifying numeric limits is to do call `std::numeric_limits<unsigned short>::max()` instead of using the C macro `USHRT_MAX`.  Honestly, the C++ approach just ends up returning the C macro anyway, but it is a more modern C++ idiom.

Sorry I didn't get these comments to you earlier.  Thanks for bearing with us.

**Comment by brballer** [comment]
Kyle,

It was a mistake to create a new module ClusterAnaV2 instead of just over-writing ClusterAna. At the time I was worried that someone might be using ClusterAna and would be negatively affected. I later spammed the larsoft distribution list asking if there were any users and didn’t get any response. I don’t have a problem replacing ClusterAna with ClusterAnaV2 but I was worried about unintended consequences. In past situations like this when there are major code changes I have gotten numerous merge conflicts with horribly mangled code.

The commented out code was left in place as a guide to future TrajCluster developers of possibly useful approaches. These were typically developed to solve a problem and significant time was spent to develop the solution, but the problem was eventually obviated by other solutions. It seems unlikely that there will be future developers other than myself so I will remove the commented code.

The std::couts are there to highlight significant problems that should be addressed. I recognize that they might not be significant to others however. I will most likely just delete them.

One could do a global search and replace of the C macro limit variables with the std versions but I don’t see the point. You need to type a lot more characters to achieve the same result.

Bruce

**Comment by knoepfel** [comment]
@brballer, thanks for the response.  At this point I suggest:

- Leave `ClusterAnaV2` as is, and we'll talk with Erica about a migration path from `ClusterAna`
- Remove commented-out code
- Your call regarding`std::cout`
- Leave the C macros for numeric limits

Will await your updates.

**Comment by brballer** [comment]
Kyle,

I made most of the changes (removing commented code) on my local repository, found numerous merge conflicts with larsoft v08_41_01 when testing (which I fixed). There are what I think compounding problems – larsoft/larreco on github and dunetpc products on redmine that don’t have d18 packages. (See below).

I give up. You can merge feature branch bb_TCWork2 to develop or not.

build_d18.x86_64> mrbsetenv
The working build directory is /Users/baller/dune/build_d18.x86_64
The source code directory is /Users/baller/dune/srcs
----------- check this block for errors -----------------------
Error encountered when setting up product: larsoft
ERROR: Product 'larsoft' (with qualifiers 'c7:debug'), has no v08_41_00 version (or may not exist)
ERROR: setup -B larsoft v08_41_00 -q +c7:+debug failed
ERROR: setup of required products has failed
build_d18.x86_64> open ../srcs/dunetpc/ups/product_deps
build_d18.x86_64> mrbsetenv
The working build directory is /Users/baller/dune/build_d18.x86_64
The source code directory is /Users/baller/dune/srcs
----------- check this block for errors -----------------------
ERROR: Action parsing failed on "unsetuprequired(dunepdsprce v1_1_1 -q +c7:+gen:+debug)"
ERROR: Found no match for product 'dunepdsprce'
ERROR: Action parsing failed on "setupRequired( dunepdsprce v1_1_1 -q +c7:+gen:+debug )"
Error encountered when setting up product: larsim
ERROR: Product 'larsim' (with qualifiers 'c7:debug'), has no v08_19_00 version (or may not exist)
ERROR: setup -B larsim v08_19_00 -q +c7:+debug failed
ERROR: setup of required products has failed

Bruce

**Comment by gartung** [comment]
Sorry looks like you have to test after Lynn publishes the v08_41_00 release (hopefully soonish)

**Comment by brballer** [comment]
Hi Bruce,

I will test things with e19 when I am back from CERN.

Thanks,
Tingjun

**Comment by gartung** [comment]
trigger build

**Comment by lgarren** [comment]
Thanks @gartung 

I am a bit confused though, since the last message I saw from Bruce said he wanted to make more changes.

**Comment by gartung** [comment]
I just sent a message to him asking if he had more changes.

**Comment by gartung** [comment]
I force pushed the old branch head without my changes.

**Comment by gartung** [comment]
At least now we know that we can make changes to the from branch and push them to the forked repo.

**Comment by gartung** [comment]
And now I put them back.

**Comment by gartung** [comment]
trigger build

**Comment by lgarren** [comment]
trigger build

Doing this again since the experiment code now matches larsoft.

**Comment by gartung** [comment]
trigger build

**Comment by gartung** [comment]
abort

**Comment by gartung** [comment]
trigger build

**Comment by lgarren** [comment]
trigger build

**Comment by gartung** [comment]
trigger build

**Comment by gartung** [comment]
trigger build

**Comment by yangtj207** [comment]
trigger build

**Comment by yangtj207** [comment]
trigger build

**Comment by yangtj207** [comment]
+1

**Comment by lgarren** [comment]
approve

**Pull request merged**